fs: improve error performance for `fdatasyncSync`

pluris opened the pull request:
```
                                                      confidence improvement accuracy (*)   (**)   (***)
fs/bench_fdatasyncSync.js n=10000 type='existing'            ***      6.39 %       ±1.85% ±2.48%  ±3.25%
fs/bench_fdatasyncSync.js n=10000 type='non-existing'        ***    116.11 %       ±6.89% ±9.21% ±12.08%

Be aware that when doing many comparisons the risk of a false-positive result increases.
In this case, there are 2 comparisons, you can thus expect the following amount of false-positive results:
  0.10 false positives, when considering a   5% risk acceptance (*, **, ***),
  0.02 false positives, when considering a   1% risk acceptance (**, ***),
  0.00 false positives, when considering a 0.1% risk acceptance (***)
```

Refs: https://github.com/nodejs/performance/issues/106
<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

pluris commented on src/node_file.cc:
```diff
@@ -116,6 +116,24 @@ inline int64_t GetOffset(Local<Value> value) {
   return IsSafeJsInt(value) ? value.As<Integer>()->Value() : -1;
 }
 
+inline int GetValidatedFd(Environment* env, Local<Value> value) {
```
Yes, if #49880 is merged, I will rebase and remove this function.

anonrig replied:
Would you mind opening a separate pull request to use valitedFd function in all existing sync functions to C++

pluris replied:
Sure, I will upload it as a separate pull request. 😄

joyeecheung reviewed (commented):
I don't think we should introduce a seemingly new implementation by copying and repeating the original implementation & moving code to a new JS file which makes git blame harder to use. See https://github.com/nodejs/node/pull/49593#issuecomment-1738089943

anonrig commented:
Hey @pluris can you rebase this pull request? You'll see that `sync.js` file is now removed and merged into `lib/fs.js`

pluris commented:
@anonrig I rebased and applied #49913.

anonrig commented:
> @anonrig I rebased and applied #49913.

Can you re-add the benchmark?

pluris commented:
@anonrig I updated the description.
```
                                                      confidence improvement accuracy (*)   (**)   (***)
fs/bench_fdatasyncSync.js n=10000 type='existing'                     1.12 %       ±1.51% ±2.01%  ±2.62%
fs/bench_fdatasyncSync.js n=10000 type='non-existing'        ***    109.42 %       ±6.13% ±8.19% ±10.73%

Be aware that when doing many comparisons the risk of a false-positive result increases.
In this case, there are 2 comparisons, you can thus expect the following amount of false-positive results:
  0.10 false positives, when considering a   5% risk acceptance (*, **, ***),
  0.02 false positives, when considering a   1% risk acceptance (**, ***),
  0.00 false positives, when considering a 0.1% risk acceptance (***)
```

anonrig commented:
The benchmark you're running doesn't exist in main or in your pull request. Can you add the file to your pull request? :-) @pluris

pluris commented:
@anonrig Oh, sorry. I added it again.
I think I did something wrong during the rebase process. 😅

pluris commented:
Oops, there was a part where the condition check was missing, so I modified it again and updated the benchmark.
```
                                                      confidence improvement accuracy (*)    (**)   (***)
fs/bench_fdatasyncSync.js n=10000 type='existing'            ***      3.68 %       ±1.72%  ±2.28%  ±2.97%
fs/bench_fdatasyncSync.js n=10000 type='non-existing'        ***    130.25 %      ±13.67% ±18.27% ±23.94%

Be aware that when doing many comparisons the risk of a false-positive result increases.
In this case, there are 2 comparisons, you can thus expect the following amount of false-positive results:
  0.10 false positives, when considering a   5% risk acceptance (*, **, ***),
  0.02 false positives, when considering a   1% risk acceptance (**, ***),
  0.00 false positives, when considering a 0.1% risk acceptance (***)
```

pluris commented:
Oops, there was a part where the condition check was missing, so I modified it again and updated the benchmark.

pluris commented:
@joyeecheung As commented in https://github.com/nodejs/node/pull/49880#discussion_r1342706699, `GetValidatedFd()` has also been removed here.
